### PR TITLE
src: lib: lore: Fix duplication when listing patchsets

### LINKED ISSUE #914: `kw patch-hub`: Fix duplication when listing patchsets
**Description**

`kw patch-hub` has duplicated patchsets when listing patchsets fetched from lore. This happens both with a query based on a string and when listing latest patchsets.

**How to Reproduce:**

*Method 1:*
1. Run `kw patch-hub`
2. Go to 'Search in Lore'
3. Type and enter a string to be searched
4. Duplicated patchsets will possibly appear next to each other

*Method 2:*
1. Run `kw patch-hub`
2. Register the `all` mailing list in 'Settings' -> 'Register/Unregister Mailing Lists'
3. Go to 'Registered mailin lists' and select the `all` list
4. Duplicated patchsets will possibly appear next to each other

**Expected behavior:**

Except when the same patchset is sent twice (virtually, they are two different patchsets with two different Message-IDs), we can detect duplicated patchsets fairly easy, so we should not pollute the interface with those.

**Screenshots:**

*Method 1:*
Image

*Method 2:* Patchsets 6 and 7
Image

**System information:**

 - OS: Manjaro kernel 6.3.13-2-MANJARO
 - kw Version: unstable branch at commit 17ebaf6b846bbf01bff31997b31ca215112338fb
 - Bash Version [use `bash --version`]

### PR DESCRIPTION
**Pull request by JGBSouza**
Listing patches off any mailing list could result in duplicate patches being displayed unnecessarily.

In this sense, solve the problem by checking if the patch already has been processed.

This PR resolves #914

### CODE REVIEW

**Comment by davidbtadokoro** [comment]
Just for the record, I had an offline conversation with @JGBSouza, and I proposed some changes to his commit. The altered commit is below.

```
diff --git a/src/lib/lore.sh b/src/lib/lore.sh
index ba17d94..f38ff4b 100644
--- a/src/lib/lore.sh
+++ b/src/lib/lore.sh
@@ -63,6 +63,10 @@ declare -g MIN_INDEX=0
 # it easy to undertand.
 declare -ag list_of_mailinglist_patches
 
+# This associative array stores the current processed patchsets and it is used
+# to check if a given patchset was already processed.
+declare -Ag processed_patchsets
+
 # This function creates the directory used by kw for any lore related data.
 #
 # Return:
@@ -299,6 +303,8 @@ function reset_current_lore_fetch_session()
   list_of_mailinglist_patches=()
   PATCHSETS_PROCESSED=0
   MIN_INDEX=0
+  unset processed_patchsets
+  declare -Ag processed_patchsets
 }
 
 # This function composes a query URL to a public mailing list archived
@@ -424,13 +430,14 @@ function process_patchsets()
     if [[ "$line" =~ ^[[:space:]]href= ]]; then
       patch_url=$(str_get_value_under_double_quotes "$line")
 
-      if is_introduction_patch "$patch_url"; then
-        # Process each patch in parallel
+      if [[ "${processed_patchsets["$patch_url"]}" != 1 ]] && is_introduction_patch "$patch_url"; then
+        # Processes each patch in parallel
         thread_for_process_patch "$PATCHSETS_PROCESSED" "$shared_dir_for_parallelism" \
           "$processed_patchset" "$patch_url" "$patch_title" &
         pids[i]="$!"
         ((i++))
         ((PATCHSETS_PROCESSED++))
+        processed_patchsets["$patch_url"]=1
       fi
 
       processed_patchset=''
diff --git a/tests/lib/lore_test.sh b/tests/lib/lore_test.sh
index 7ea4a16..97aa61c 100755
--- a/tests/lib/lore_test.sh
+++ b/tests/lib/lore_test.sh
@@ -559,11 +559,16 @@ function test_reset_current_lore_fetch_session()
   list_of_mailinglist_patches[2]=1
   PATCHSETS_PROCESSED=3
   MIN_INDEX=200
+  declare -Ag processed_patchsets
+  processed_patchsets['patch_1']=1
+  processed_patchsets['patch_2']=1
+  processed_patchsets['patch_3']=1
 
   reset_current_lore_fetch_session 2
   assert_equals_helper 'Should reset `list_of_mailinglist_patches`' "$LINENO" 0 "${#list_of_mailinglist_patches[@]}"
   assert_equals_helper 'Should reset `PATCHSETS_PROCESSED`' "$LINENO" 0 "$PATCHSETS_PROCESSED"
   assert_equals_helper 'Should reset `MIN_INDEX`' "$LINENO" 0 "$MIN_INDEX"
+  assert_equals_helper 'Should reset `processed_patchsets`' "$LINENO" 0 "${#processed_patchsets[@]}"
 }
 
 function test_format_patchsets()
```

**Comment by JGBSouza** [comment]
Hi @davidbtadokoro, thanks for the updates. Looking to all the changes they really seem to make the code better

**Comment by davidbtadokoro** [comment]
@rodrigosiqueira @JGBSouza I came upon a possible concurrency problem. Because we process each patch in parallel, it is possible for a thread to pass the check if the patchset was processed, when it shouldn't.

For example, imagine thread 1 gets patchset of  message ID `foo`. Thread 1 enters the if statement and goes into a background to process the patchset. Until it finishes the process, assign `processed_patchsets["$patch_url"]=1` does not happen. Suppose thread 2 also gets patchset of  message ID `foo`, the thread will also enter the if statement and process the same patchset.

All this to say, do you think that this makes sense? Think I just stumbled into this behavior, but didn't screenshot 😞. Below I put the part of the code that I think the concurrency problem may happen, and added the two points I think the assign `processed_patchsets["$patch_url"]=1` should happen (the first being the one that prevents concurrency and the second being the "correct" one). What do you guys think?

```bash
-      if is_introduction_patch "$patch_url"; then
-        # Process each patch in parallel
+      if [[ "${processed_patchsets["$patch_url"]}" != 1 ]] && is_introduction_patch "$patch_url"; then
+        processed_patchsets["$patch_url"]=1 # <--- LOCK AT THE BEGINNING
+        # Processes each patch in parallel
         thread_for_process_patch "$PATCHSETS_PROCESSED" "$shared_dir_for_parallelism" \
           "$processed_patchset" "$patch_url" "$patch_title" &
         pids[i]="$!"
         ((i++))
         ((PATCHSETS_PROCESSED++))
+        processed_patchsets["$patch_url"]=1 # <--- LOCK AT THE END
       fi
```

**Comment by davidbtadokoro** [comment]
> @rodrigosiqueira @JGBSouza I came upon a possible concurrency problem. Because we process each patch in parallel, it is possible for a thread to pass the check if the patchset was processed, when it shouldn't.

After offline discussion I realized that my comment did not make sense.

File: tests/unit/lib/lore_test.sh
Comment (rodrigosiqueira) [inline tests/unit/lib/lore_test.sh:1]: Just a note about the commit message: if Joao helped to develop this change, you should not add his reviewed-by. Use Co-developed-by instead.

File: tests/unit/lib/lore_test.sh [hunk 1]
```diff
@@ -510,52 +510,57 @@ function test_pre_process_xml_result()
...
+  # Process first batch of pre processed patches
+  pre_processed_patches_sample=$(< "${SHUNIT_TMPDIR}/samples/pre_processed_patches_sample-2")
+  process_patchsets "$pre_processed_patches_sample"
+  assert_equals_helper 'Wrong number of patchsets processed' "$LINENO" 3 "$PATCHSETS_PROCESSED"
+  expected='Gilberto GilÆgil.gil@mpb.brÆXÆXÆ[PATCH v3] Add Palco to MPBÆhttp://lore.kernel.org/mpb/introduction'
+  assert_equals_helper 'Wrong processed patchset (index 0)' "$LINENO" "$expected" "${list_of_mailinglist_patches[0]}"
+  expected='David BowieÆmajor.tom@rock.ukÆXÆXÆ[RFC PATCH v12] Introduce Ziggy StardustÆhttp://lore.kernel.org/rock/introduction'
+  assert_equals_helper 'Wrong processed patchset (index 1)' "$LINENO" "$expected" "${list_of_mailinglist_patches[1]}"
+  expected='Seu JorgeÆseu.jorge@samba-pop.brÆXÆXÆ[RFC] Release Músicas para Churrasco Vol.1Æhttp://lore.kernel.org/samba-pop/introduction'
+  assert_equals_helper 'Wrong processed patchset (index 2)' "$LINENO" "$expected" "${list_of_mailinglist_patches[2]}"
```
Comment (rodrigosiqueira) [inline tests/unit/lib/lore_test.sh:563]: Add new lines in the `assert_equals_helper` areas for readability's sake. Maybe consider splitting these checks into two or more test functions?

File: tests/unit/lib/lore_test.sh [hunk 2]
```diff
@@ -563,18 +563,40 @@ function test_process_patchsets()
...
+  local expected
+
+  # Clear number of patchsets processed and data structure with patchsets
+  reset_current_lore_fetch_session
+
+  # Process list of repeated pre processed patches
+  pre_processed_patches_sample=$(< "${SHUNIT_TMPDIR}/samples/pre_processed_patches_sample-repeated")
+  process_patchsets "$pre_processed_patches_sample"
+  assert_equals_helper 'Repeated patches should not be processed again' "$LINENO" 1 "$PATCHSETS_PROCESSED"
+  expected='David BowieÆmajor.tom@rock.ukÆXÆXÆ[RFC PATCH v12] Introduce Ziggy StardustÆhttp://lore.kernel.org/rock/introduction'
```
Comment (rodrigosiqueira) [inline tests/unit/lib/lore_test.sh:578]: Add a new line

File: tests/unit/lib/lore_test.sh [hunk 3]
```diff
@@ -563,18 +563,40 @@ function test_process_patchsets()
...
+  assert_equals_helper 'Wrong processed patchset (index 0)' "$LINENO" "$expected" "${list_of_mailinglist_patches[0]}"
+
```
Comment (rodrigosiqueira) [inline tests/unit/lib/lore_test.sh:580]: Drop this extra line.

File: tests/unit/lib/lore_test.sh [hunk 4]
```diff
@@ -563,18 +563,40 @@ function test_process_patchsets()
...
 function test_reset_current_lore_fetch_session()
 {
   list_of_mailinglist_patches[0]=1
   list_of_mailinglist_patches[1]=1
   list_of_mailinglist_patches[2]=1
   PATCHSETS_PROCESSED=3
   MIN_INDEX=200
+  declare -Ag processed_patchsets
+  processed_patchsets['patch_1']=1
```
Comment (rodrigosiqueira) [inline tests/unit/lib/lore_test.sh:591]: Add new line.

**Review by rodrigosiqueira — CHANGES_REQUESTED** [review]

**Comment by rodrigosiqueira** [comment]
Hi @JGBSouza , please try to update this PR.

**Comment by davidbtadokoro** [comment]
Thanks so much for the continuing working in this PR @JGBSouza! It fixed the issue and it looks legit, so merged into the unstable branch :+1:

**Pull request closed without merging**